### PR TITLE
Streaming store operation for VectorizedArray

### PR DESCRIPTION
**Pull request by kronbichler**
See the changelog for a description of this patch.

### CODE REVIEW

**Comment by kronbichler** [comment]
Regarding the name of the new function: I chose "store_bypassing_caches" because it is probably most intuitive to users, whereas the official Intel documents call it "streaming store". At least I did not know what that meant before looking it up. What do others think?

**Comment by masterleinad** [comment]
Honestly, I would just go with `streaming_store` as this matches the technical term that can be found in documentations closely. People not knowing about `streaming stores` would probably have to look this up anyway.

**Comment by kronbichler** [comment]
Thinking about it, I definitely agree. It only helps users who know about those things, and `streaming_store` is what exists in literature. I have change the name to `streaming_store`.

File: include/deal.II/base/vectorization.h [hunk 1]
```diff
@@ -262,6 +262,26 @@ class VectorizedArray
     *ptr = data;
   }
 
+  /**
+   * Write the content of the calling class into memory in form of @p
+   * n_array_elements to the given address using non-temporal stores that
+   * bypass the processor's caches, using _mm_stream_pd store intrinsics on
+   * supported CPUs. The memory must be aligned by the amount of bytes in the
```
Comment (bangerth) [inline include/deal.II/base/vectorization.h]: Do you mean the *destination* of the store?

File: include/deal.II/base/vectorization.h [hunk 2]
```diff
@@ -262,6 +262,26 @@ class VectorizedArray
...
+   * vectorized array. This store operation can be faster than the other store
+   * operations in case the store is streaming because it avoids the
+   * read-for-ownership transfer typically invoked in standard stores that
+   * must first read in the data to be written. Since this operation does not
+   * use caches, it typically takes more time to read it again after writing.
```
Comment (bangerth) [inline include/deal.II/base/vectorization.h]: can you expand on what "it" refers to here in both places?

File: include/deal.II/base/vectorization.h [hunk 3]
```diff
@@ -262,6 +262,26 @@ class VectorizedArray
     *ptr = data;
   }
 
+  /**
+   * Write the content of the calling class into memory in form of @p
+   * n_array_elements to the given address using non-temporal stores that
+   * bypass the processor's caches, using _mm_stream_pd store intrinsics on
```
Comment (bangerth) [inline include/deal.II/base/vectorization.h]: -> @p _mm_stream_pd

File: include/deal.II/base/vectorization.h [hunk 4]
```diff
@@ -628,6 +648,24 @@ class VectorizedArray<double>
     _mm512_storeu_pd (ptr, data);
   }
 
+  /**
+   * Write the content of the calling class into memory in form of @p
```
Comment (bangerth) [inline include/deal.II/base/vectorization.h]: same comments as above.

File: include/deal.II/base/vectorization.h [hunk 5]
```diff
@@ -1008,6 +1046,24 @@ class VectorizedArray<float>
     _mm512_storeu_ps (ptr, data);
   }
 
+  /**
+   * Write the content of the calling class into memory in form of @p
```
Comment (bangerth) [inline include/deal.II/base/vectorization.h]: and here

File: include/deal.II/base/vectorization.h [hunk 6]
```diff
@@ -1414,6 +1470,24 @@ class VectorizedArray<double>
     _mm256_storeu_pd (ptr, data);
   }
 
+  /**
+   * Write the content of the calling class into memory in form of @p
```
Comment (bangerth) [inline include/deal.II/base/vectorization.h]: and here, and elsewhere as well, of course.

**Comment by kronbichler** [comment]
@bangerth Thanks for the comments. I extended the documentation of the new function by covering the (in my opinion) essential components of it. In order to avoid copying almost 50 lines to 6 other places, I used `@copydoc` now.

**Comment by bangerth** [comment]
And this is why deal.II has a reputation for being very well documented! Thanks for writing all of this up!

In a few hours, take a look at the documentation to make sure the `@copydoc` actually works (unless, of course, you double checked this locally already).

**Pull request merged**